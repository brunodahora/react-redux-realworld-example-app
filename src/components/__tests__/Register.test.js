import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { ConnectedRouter } from 'react-router-redux';
import { Provider } from "react-redux";
import { store, history } from "../../store";
import Register from "../Register";

describe("<Register />", () => {
  describe("Given that I'm in the Sign up page", () => {
    beforeEach(() => {
      render(
        <Provider store={store}>
          <ConnectedRouter history={history}>
            <Register />
          </ConnectedRouter>
        </Provider>
      );
    });

    describe("when I have inserted a valid username", () => {
      beforeEach(() => {
        fireEvent.change(screen.getByPlaceholderText(/username/i), { target: { value: 'test08102020' } });
      });

      describe("and clicking sign up", () => {
        beforeEach(() => {
          fireEvent.click(screen.getByRole("button", { name: "Sign up" }));
        });

        test("then the email can't be blank error is shown", async () => {
          await waitFor(() => expect(screen.getByText("email can't be blank")).toBeInTheDocument());
        });

        test("then the password can't be blank error is shown", async () => {
          await waitFor(() => expect(screen.getByText("password can't be blank")).toBeInTheDocument());
        });
      });

      describe("and I have inserted an invalid email", () => {
        beforeEach(() => {
          fireEvent.change(screen.getByPlaceholderText(/email/i), { target: { value: 'test@test' } });
        });

        describe("and clicking sign up", () => {
          beforeEach(() => {
            fireEvent.click(screen.getByRole("button", { name: "Sign up" }));
          });
  
          test("then the email is invalid error is shown", async () => {
            await waitFor(() => expect(screen.getByText("email is invalid")).toBeInTheDocument());
          });
  
          test("then the password can't be blank error is shown", async () => {
            await waitFor(() => expect(screen.getByText("password can't be blank")).toBeInTheDocument());
          });
        });
      });
      
      describe("and I have inserted a valid email", () => {
        beforeEach(() => {
          fireEvent.change(screen.getByPlaceholderText(/email/i), { target: { value: 'testbruno2@test.com' } });
        });

        describe("and clicking sign up", () => {
          beforeEach(() => {
            fireEvent.click(screen.getByRole("button", { name: "Sign up" }));
          });

          test("then the password can't be blank error is shown", async () => {
            await waitFor(() => expect(screen.getByText("password can't be blank")).toBeInTheDocument());
          });
        });

        describe("and I have inserted a short password", () => {
          beforeEach(() => {
            fireEvent.change(screen.getByPlaceholderText(/password/i), { target: { value: '123456' } });
          });

          describe("and clicking sign up", () => {
            beforeEach(() => {
              fireEvent.click(screen.getByRole("button", { name: "Sign up" }));
            });
  
            test("then the password is too short error is shown", async () => {
              await waitFor(() => expect(screen.getByText("password is too short (minimum is 8 characters)")).toBeInTheDocument());
            });
          });
        });

        describe("and I have inserted a good password", () => {
          beforeEach(() => {
            fireEvent.change(screen.getByPlaceholderText(/password/i), { target: { value: '123456789' } });
          });

          describe("and clicking sign up", () => {
            beforeEach(() => {
              fireEvent.click(screen.getByRole("button", { name: "Sign up" }));
            });
  
            test("then I should be redirected to the main page", async () => {
              // check if redirected
            });
          });
        });
      });
    });
  });
});
