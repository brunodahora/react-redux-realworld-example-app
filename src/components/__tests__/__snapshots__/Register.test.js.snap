// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> should render 1`] = `
<div
  class="auth-page"
>
  <div
    class="container page"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 offset-md-3 col-xs-12"
      >
        <h1
          class="text-xs-center"
        >
          Sign Up
        </h1>
        <p
          class="text-xs-center"
        >
          <a
            href="/login"
          >
            Have an account?
          </a>
        </p>
        <form>
          <fieldset>
            <fieldset
              class="form-group"
            >
              <input
                class="form-control form-control-lg"
                placeholder="Username"
                type="text"
                value=""
              />
            </fieldset>
            <fieldset
              class="form-group"
            >
              <input
                class="form-control form-control-lg"
                placeholder="Email"
                type="email"
                value=""
              />
            </fieldset>
            <fieldset
              class="form-group"
            >
              <input
                class="form-control form-control-lg"
                placeholder="Password"
                type="password"
                value=""
              />
            </fieldset>
            <button
              class="btn btn-lg btn-primary pull-xs-right"
              type="submit"
            >
              Sign up
            </button>
          </fieldset>
        </form>
      </div>
    </div>
  </div>
</div>
`;
